dkc/core/rest/file.py

```diff
@@ -109,3 +113,27 @@ def download(self, request, pk=None):
         if file.blob:  # FieldFiles are falsy when not populated with a file
             return HttpResponseRedirect(file.blob.url)
         return Response(status=204)
+
+    @swagger_auto_schema(
+        query_serializer=HashDownloadSerializer,
+        responses={
+            200: None,  # needed to override default
+            302: 'You will be redirected to download the file contents.',
+            404: 'No file exists with the given hash.',
+        },
+    )
+    @action(detail=False, filterset_fields=[], pagination_class=None)
+    def hash_download(self, request):
+        """Download a file based on the hash of its contents."""
+        serializer = HashDownloadSerializer(data=request.query_params)
+        serializer.is_valid(raise_exception=True)
+        sha512 = serializer.validated_data['sha512'].lower()
+        qs = File.objects.filter(sha512=sha512).only('blob').order_by()
+        qs = File.filter_by_permission(request.user, Permission.read, qs)
+
+        try:
+            file = qs[:1].get()
+        except File.DoesNotExist:
+            return Response(status=404)
```

Comment:
```suggestion
        file = qs.first()
        if not file:
            return Response(status=404)
```

https://docs.djangoproject.com/en/3.1/ref/models/querysets/#first seems more straightforward.

Reply:
`first` is exactly the behavior I want. Thanks for the tip -- I was reading lots of Django docs for this case and somehow didn't find it.

Reply:
It turns out, `qs.first()` has a different query output than `qs[:1].get()`. The difference is, `first` adds an `ORDER BY "core_file"."id" ASC` that is not present in the `.get` version. This is [documented behavior](https://docs.djangoproject.com/en/3.1/ref/models/querysets/#django.db.models.query.QuerySet.first):

> If the QuerySet has no ordering defined, then the queryset is automatically ordered by the primary key.

The implementation is:

```python
def first(self):
        """Return the first object of a query or None if no match is found."""
        for obj in (self if self.ordered else self.order_by('pk'))[:1]:
            return obj
```

I don't think that's very sensible behavior, but there it is. In light of that I'm inclined to go back to the old way, just because I went out of the way to tell the database that no sorting was necessary.

Reply:
Then again, I suspect that because the PK is the clustering index, it won't incur a real performance penalty in practice, it's just irksome that I can't use `first` the way I want.